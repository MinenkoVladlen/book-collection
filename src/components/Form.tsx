import React from "react";
import { Button, TextField } from "@mui/material";
import { useFormik } from "formik";
import booksStore from "../services/books/Books.store.ts";
import * as yup from "yup";
import '../styles.css';

const validationSchema = yup.object({
  name: yup.string().required("Book name is required"),
  author: yup.string().required("Author is required")
});


const Form: React.FC = () => {
    const formik = useFormik({
        initialValues: {
          name: "",
          author: ""
        },
        validationSchema: validationSchema,
        onSubmit: async (values) => {
          await booksStore.addBook(values);
          formik.resetForm();
        }
      });

    return (
        <form onSubmit={formik.handleSubmit} id='Form' data-testid="form">
            <TextField
                label="Book Name"
                name="name"
                value={formik.values.name}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                error={formik.touched.name && Boolean(formik.errors.name)}
                helperText={formik.touched.name && formik.errors.name}
                margin="normal"
            />
            <TextField
                label="Author"
                name="author"
                value={formik.values.author}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                error={formik.touched.author && Boolean(formik.errors.author)}
                helperText={formik.touched.author && formik.errors.author}
                margin="normal"
            />
            <Button variant="contained" type="submit" sx={{margin: 'auto'}}>Add</Button>
        </form>
    )
}

export default Form;