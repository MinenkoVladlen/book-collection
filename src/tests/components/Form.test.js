import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Form from '../../components/Form';
import booksStore from '../../services/books/Books.store';

jest.mock('../../services/books/Books.store');

describe('Form', () => {
  beforeEach(() => {
    booksStore.books = [];
  });

    it('should render correctly', () => {
        const { asFragment } = render(<Form />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should handle form submission', () => {
        const { getByTestId } = render(<Form />);
        const form = getByTestId('form');
        const handleSubmit = jest.fn();

        form.onsubmit = handleSubmit;
        form.dispatchEvent(new Event('submit'));

        expect(handleSubmit).toHaveBeenCalled();
    });
});
