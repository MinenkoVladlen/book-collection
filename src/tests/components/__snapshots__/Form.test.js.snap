// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render correctly 1`] = `
<DocumentFragment>
  <form
    data-testid="form"
    id="Form"
  >
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1fw0kwd-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Book Name
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          name="name"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-81qg8w"
          >
            <span>
              Book Name
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1fw0kwd-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r1:"
        id=":r1:-label"
      >
        Author
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r1:"
          name="author"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-81qg8w"
          >
            <span>
              Author
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1oywgji-MuiButtonBase-root-MuiButton-root"
      id=":r2:"
      tabindex="0"
      type="submit"
    >
      Add
    </button>
  </form>
</DocumentFragment>
`;
